### PR TITLE
[Wallet] Upgrade detox, react-native, hermes in order to fix common crashes

### LINKED ISSUE #6104: [P0] Fix e2e tests
# What's important about this?
* We have a probably serious issue in the production, which we can reproduce on e2e test (Android)
# What is "done"?
* E2e are passing on both platforms without flakiness
# What's involved in doing this work?
* Fix flakiness of end-to-end-mobile-test-android and end-to-end-mobile-test-ios
  * Run in it release mode in Android
  * Resolve detox upgrade issue

### PR DESCRIPTION
**Pull request by etuleu**
### Description

This PR is aimed at fixing end to end tests as well as fixing some of the top crashes.

Updated the following core packages:
"react": "16.13.1",
"react-native": "^0.63.4",
"detox": "^17.13.2",
"react-native-reanimated": "^2.0.0-rc.0",
"@react-navigation/compat": "^5.3.10",
"@react-navigation/drawer": "^5.11.4",
"@react-navigation/material-top-tabs": "^5.3.10",
"@react-navigation/native": "^5.8.10",
"@react-navigation/stack": "^5.12.8",

On the android side:
gradle plugin 3.5.3

Sorry for the length of the PR. Some things you can ignore are the test snapshots. I preferred having a longer PR that builds and passes tests than a bunch of smaller PRs that depend on each other.

I think the e2e tests were actually exposing a serious problem, which happened on app start. I started by upgrading only detox and then figuring out what else was needed. With just detox upgraded to the latest the app started consistently crashing, with similar traces to the ones we see in prod. I believe this is related to us running an older hermes-engine, so I upgraded that as well. The benefit now is that the crashes also look much more readable which was a feature added recently by hermes (they were lacking native symbols before).

The PR got a little big, but I believe the changes all go together since the packages are inter related. I would really appreciate you starting to take a look @jeanregisser and @cmcewen since you've seen some of this before.

### Other changes

The changes don't seem related to each other but they all seemed necessary in order to silence the type errors.

### Tested

Manually and e2e.

### Related issues

- Fixes #6104 #6200 #6201 #6202 #6204 #6205

### Backwards compatibility

Yes.

### CODE REVIEW

File: packages/flake-tracker/package.json [hunk 1]
```diff
@@ -18,6 +18,8 @@
     "@octokit/plugin-retry": "^3.0.3",
     "@octokit/rest": "^18.0.0",
     "clone": "^2.1.2",
+    "detox": "^17.13.2",
+    "mocha": "^7.1.1",
```
Comment (jeanregisser) [inline packages/flake-tracker/package.json:22]: Why is this needed in the `flake-tracker` package?

File: packages/mobile/e2e/src/usecases/HandleDeepLinkDappkit.js [hunk 2]
```diff
@@ -1,6 +1,10 @@
...
+    device.getPlatform() === 'ios'
+      ? 'celo://wallet/dappkit?type=sign_tx&requestId=beneficiaryclaim&callback=impactmarket%3A%2F%2F&dappName=impactmarket&txs=W3sidHhEYXRhIjoiMHg0ZTcxZDkyZCIsImVzdGltYXRlZEdhcyI6MTIxMTI3LCJmcm9tIjoiMHhhY2FGQjRGMWQ5RERGMTQwNWViZEJGZWM5NTlBNzI4MTk0QjAyMzhhIiwidG8iOiIweDc0ZTVDNDA1RURFNEUzN2U3ODBEQjk2NTI1NDE1MzhkZDhBNzlBN2QiLCJub25jZSI6MjYsImZlZUN1cnJlbmN5QWRkcmVzcyI6IjB4NzY1REU4MTY4NDU4NjFlNzVBMjVmQ0ExMjJiYjY4OThCOEIxMjgyYSIsInZhbHVlIjoiMCJ9XQ%3D%3D'
+      : '"celo://wallet/dappkit?type=sign_tx&requestId=beneficiaryclaim&callback=impactmarket%3A%2F%2F&dappName=impactmarket&txs=W3sidHhEYXRhIjoiMHg0ZTcxZDkyZCIsImVzdGltYXRlZEdhcyI6MTIxMTI3LCJmcm9tIjoiMHhhY2FGQjRGMWQ5RERGMTQwNWViZEJGZWM5NTlBNzI4MTk0QjAyMzhhIiwidG8iOiIweDc0ZTVDNDA1RURFNEUzN2U3ODBEQjk2NTI1NDE1MzhkZDhBNzlBN2QiLCJub25jZSI6MjYsImZlZUN1cnJlbmN5QWRkcmVzcyI6IjB4NzY1REU4MTY4NDU4NjFlNzVBMjVmQ0ExMjJiYjY4OThCOEIxMjgyYSIsInZhbHVlIjoiMCJ9XQ%3D%3D"'
```
Comment (jeanregisser) [inline packages/mobile/e2e/src/usecases/HandleDeepLinkDappkit.js]: Let's avoid the repetition of the content and maybe introduce a little helper function which only takes care of adding the the double quotes on Andrdoid?
I see this is needed in another place too.

File: packages/mobile/package.json [hunk 3]
```diff
@@ -119,7 +119,7 @@
     "react-native-progress": "^4.1.2",
     "react-native-qrcode-svg": "^6.0.6",
     "react-native-randombytes": "^3.6.0",
-    "react-native-reanimated": "^1.13.1",
+    "react-native-reanimated": "^2.0.0-rc.0",
```
Comment (jeanregisser) [inline packages/mobile/package.json]: `reanimated 2` introduces big changes, no hidden breakage with this?

File: packages/mobile/package.json [hunk 4]
```diff
@@ -161,29 +161,30 @@
...
     "escape-string-regexp": "^1.0.5",
     "eslint": "^6.8.0",
     "jest-fetch-mock": "^3.0.3",
-    "metro-react-native-babel-preset": "^0.58.0",
+    "metro-react-native-babel-preset": "^0.59.0",
+    "mocha": "^7.1.1",
     "patch-package": "^5.1.1",
     "postinstall-prepare": "^1.0.1",
-    "react-dom": "16.11.0",
+    "react-dom": "16.13.1",
```
Comment (jeanregisser) [inline packages/mobile/package.json]: Surprised we have this package. I don't see why we added it. Let's remove it?

File: packages/mobile/package.json [hunk 5]
```diff
@@ -161,29 +161,30 @@
...
     "@types/utf8": "^2.1.6",
     "babel-core": "7.0.0-bridge.0",
-    "detox": "^17.4.5",
+    "detox": "^17.13.2",
     "escape-string-regexp": "^1.0.5",
     "eslint": "^6.8.0",
     "jest-fetch-mock": "^3.0.3",
-    "metro-react-native-babel-preset": "^0.58.0",
+    "metro-react-native-babel-preset": "^0.59.0",
+    "mocha": "^7.1.1",
```
Comment (jeanregisser) [inline packages/mobile/package.json:187]: Why is `mocha` needed?

File: packages/mobile/package.json [hunk 6]
```diff
@@ -161,29 +161,30 @@
...
-    "@react-native-community/eslint-config": "^1.1.0",
+    "@react-native-community/eslint-config": "^2.0.0",
     "@types/crypto-js": "^3.1.47",
     "@types/ethereumjs-util": "^5.2.0",
     "@types/graphql": "^14.0.7",
     "@types/isomorphic-fetch": "^0.0.35",
     "@types/lodash": "^4.14.136",
-    "@types/react": "^16.9.34",
-    "@types/react-native": "^0.62.4",
+    "@types/react": "^17.0.0",
```
Comment (jeanregisser) [inline packages/mobile/package.json]: Might be harmless but we shouldn't use the definitions for `react 17.x` since the project still uses `react 16.x`

File: packages/mobile/package.json [hunk 7]
```diff
@@ -161,29 +161,30 @@
     "@graphql-codegen/fragment-matcher": "1.10.0",
     "@graphql-codegen/typescript": "1.10.0",
     "@graphql-codegen/typescript-operations": "1.10.0",
-    "@react-native-community/eslint-config": "^1.1.0",
+    "@react-native-community/eslint-config": "^2.0.0",
```
Comment (jeanregisser) [inline packages/mobile/package.json]: A brand new RN 0.63.4 project still recommends: `"@react-native-community/eslint-config": "^1.1.0",`

```
"devDependencies": {
    "@babel/core": "^7.8.4",
    "@babel/runtime": "^7.8.4",
    "@react-native-community/eslint-config": "^1.1.0",
    "babel-jest": "^25.1.0",
    "eslint": "^6.5.1",
    "jest": "^25.1.0",
    "metro-react-native-babel-preset": "^0.59.0",
    "react-test-renderer": "16.13.1"
  },
```

File: patches/react-native-tab-view+2.14.4.patch [hunk 8]
```diff
@@ -1,36 +0,0 @@
...
-       jumpTo: this.jumpTo,
---- a/node_modules/react-native-tab-view/lib/module/ScrollPager.js
-+++ b/node_modules/react-native-tab-view/lib/module/ScrollPager.js
-@@ -155,6 +155,9 @@ export default class ScrollPager extends React.Component {
-       removeListener: this.removeListener,
-       jumpTo: this.jumpTo,
-       render: children => /*#__PURE__*/React.createElement(Animated.ScrollView, {
-+        // This component should accept ScrollView properties
-+        // for now just patch what we need
-+        alwaysBounceVertical: false,
```
Comment (jeanregisser) [inline patches/react-native-tab-view+2.14.4.patch]: Without this patch the QR scan screen can bounce vertically which doesn't look nice.

File: patches/react-native-tab-view+2.14.4.patch [hunk 9]
```diff
@@ -1,36 +0,0 @@
...
-@@ -326,6 +326,11 @@ export default class Pager extends React.Component {
- 
-     _defineProperty(this, "getTranslateX", memoize((layoutWidth, routesLength, translateX) => multiply( // Make sure that the translation doesn't exceed the bounds to prevent overscrolling
-     min(max(multiply(layoutWidth, sub(routesLength, 1), DIRECTION_RIGHT), translateX), 0), I18nManager.isRTL ? -1 : 1)));
-+
-+    // Makes sure the reported position is based on the actual translation
-+    // so our custom animation that follows it, looks good
-+    // without this, it looks like it's lagging behind the gesture
-+    _defineProperty(this, "reportedPosition", cond(this.layoutWidth, divide(multiply(this.translateX, -1), this.layoutWidth), this.index));
-   }
```
Comment (jeanregisser) [inline patches/react-native-tab-view+2.14.4.patch]: Quickly retested this on iOS and it seems it's not needed anymore. But I don't remember if this might have been needed specifically for Android.

File: packages/reserve-site/package.json [hunk 10]
```diff
@@ -22,15 +22,15 @@
     "markdown-to-jsx": "^6.11.4",
     "next": "^9.5.2",
     "raw-loader": "^4.0.1",
-    "react": "16.11.0",
-    "react-dom": "16.11.0"
+    "react": "16.13.1",
+    "react-dom": "16.13.1"
```
Comment (jeanregisser) [inline packages/reserve-site/package.json:26]: @aaronmgdr should these be left alone and not modified by this PR?

File: packages/web/package.json [hunk 11]
```diff
@@ -69,15 +69,15 @@
     "node-cache": "^5.0.2",
     "nuka-carousel": "^4.5.13",
     "raw-loader": "^4.0.1",
-    "react": "16.11.0",
+    "react": "16.13.1",
```
Comment (jeanregisser) [inline packages/web/package.json:72]: @aaronmgdr ditto

File: packages/react-components/components/KeyboardAwareScrollView.tsx [hunk 12]
```diff
@@ -1,96 +0,0 @@
...
-} from 'react-native'
-
-type ComponentOrHandle = null | number | React.Component<any, any> | React.ComponentClass<any>
-
-interface Props extends ScrollViewProps {
-  hasNavBar?: boolean
-}
-
-// Keeps the focused input into view when the keyboard is shown
-export default class KeyboardAwareScrollView extends React.Component<Props> {
```
Comment (jeanregisser) [inline packages/react-components/components/KeyboardAwareScrollView.tsx]: if possible, I'd keep that component and fix any new issue.
`react-native-keyboard-aware-scroll-view` is a bit overkill and would avoid the extra dependency.

**Review by jeanregisser — COMMENTED** [review]
Awesome work!! 👍 
Could you also change the title/description of the PR to reflect that it's larger work than just the Detox upgrade.
Also can you link existing issues which this work will fix?

File: patches/react-native-tab-view+2.14.4.patch (hunk 9 above)
Comment (jeanregisser) [reply patches/react-native-tab-view+2.14.4.patch]: I just checked on Android and this specific part of the patch indeed needed there!

File: packages/reserve-site/package.json (hunk 10 above)
Comment (aaronmgdr) [reply packages/reserve-site/package.json:26]: Actually please update them to be the same as on wallet. Ive had weird things happen when the yarn.lock has multiple versions of react.

File: packages/web/package.json (hunk 11 above)
Comment (aaronmgdr) [reply packages/web/package.json:72]: update away

File: packages/flake-tracker/package.json (hunk 1 above)
Comment (etuleu) [reply packages/flake-tracker/package.json:22]: I couldn't remember why but I just tried removing it and I get an error while building android that says it can't find com.wix.detox:

`Could not determine the dependencies of task ':app:preApptestDebugAndroidTestBuild'.
> Could not resolve all task dependencies for configuration ':app:apptestDebugAndroidTestRuntimeClasspath'.
   > Could not find any matches for com.wix:detox:+ as no versions of com.wix:detox are available.
     Searched in the following locations:
       - file:/home/erdal/.m2/repository/com/wix/detox/
       - file:/home/erdal/dev/celo-monorepo/node_modules/react-native/android/com/wix/detox/maven-metadata.xml
       - file:/home/erdal/dev/celo-monorepo/node_modules/jsc-android/dist/com/wix/detox/maven-metadata.xml
       - file:/home/erdal/dev/celo-monorepo/node_modules/detox/Detox-android/com/wix/detox/maven-metadata.xml
       - file:/home/erdal/dev/celo-monorepo/node_modules/@celo/client/build/bin/
       - https://dl.google.com/dl/android/maven2/com/wix/detox/maven-metadata.xml
       - https://jcenter.bintray.com/com/wix/detox/maven-metadata.xml
       - https://www.jitpack.io/com/wix/detox/maven-metadata.xml
     Required by:
         project :app
`

Obviously this is the wrong solution, so trying to find a better one.

Comment (etuleu) [reply packages/flake-tracker/package.json:22]: I think the problem is detox gets installed in packages/mobile/node_modules and our gradle script looks for it in the root node_modules.

Comment (etuleu) [reply packages/flake-tracker/package.json:22]: Ok this is taking me back to a few weeks ago :) When I remove the package from flake-tracker I get errors when running the e2e tests:

`detox[21699] INFO:  [test.js] configuration="android.emu.debug" loglevel="verbose" artifactsLocation="e2e/artifacts" recordLogs="failing" takeScreenshots="all" useCustomLogger=true forceAdbInstall=false DETOX_START_TIMESTAMP=1607626703988 readOnlyEmu=false reportSpecs=true jest --config e2e/config.json --testNamePattern '^((?!:ios:).)*$' --maxWorkers 1 --detectOpenHandles  
 FAIL  e2e/src/DeepLinkSend.spec.js
  ● Test suite failed to run

    Cannot find module 'detox/runners/jest-circus'

      3 |   SpecReporter,
      4 |   WorkerAssignReporter,
    > 5 | } = require('detox/runners/jest-circus')
        |     ^
      6 | const JestFlakeTracker = require('../tracker')
      7 | 
      8 | class FlakeTrackingDetoxEnv extends DetoxCircusEnvironment {

      at Object.<anonymous> (../../flake-tracker/src/jest/environments/detox.js:5:5)

`

Comment (etuleu) [reply packages/flake-tracker/package.json:22]: so I think it is needed there. What probably happens is the package gets hoisted to the root, which is why we import it from there in the gradle build:
`maven {
            url "$rootDir/../../../node_modules/detox/Detox-android"
        }`

File: packages/mobile/package.json (hunk 3 above)
Comment (etuleu) [reply packages/mobile/package.json]: Not that I have noticed but we should extensively test. Unfortunately animations are hard to test automatically so have to pay attention in this release. Anything I should manually test related that you can think of?

The reason I want to upgrade this is it fixes a lot of the crashes we have been seeing such as this one:
https://github.com/software-mansion/react-native-reanimated/pull/1457

File: packages/mobile/package.json (hunk 5 above)
Comment (etuleu) [reply packages/mobile/package.json:187]: without mocha I get this error:
`error An unexpected error occurred: "could not find a copy of mocha to link in /home/erdal/dev/celo-monorepo/node_modules/detox/node_modules".
`

File: patches/react-native-tab-view+2.14.4.patch (hunk 8 above)
Comment (etuleu) [reply patches/react-native-tab-view+2.14.4.patch]: ok bringing it back and make it a patch for the new version 2.15.2

File: packages/react-components/components/KeyboardAwareScrollView.tsx (hunk 12 above)
Comment (etuleu) [reply packages/react-components/components/KeyboardAwareScrollView.tsx]: I agree it's annoying to bring in another dependency, but considering we are using it in some critical components and it causes crashes I think we should use the latest version of this library since it seems to be used and maintained pretty well. For example they recently added fixes for RN 0.63 as well as for iPhone X. Also the library itself is pretty small so I don't think it's a big increase to the app.

https://github.com/APSL/react-native-keyboard-aware-scroll-view/releases
https://github.com/APSL/react-native-keyboard-aware-scroll-view/blob/master/lib/KeyboardAwareHOC.js

Comment (jeanregisser) [reply packages/react-components/components/KeyboardAwareScrollView.tsx]: The problem is `react-native-keyboard-aware-scroll-view` doesn't work nicely with our `KeyboardSpacer` component and breaks the code input screen for instance:



I built our `KeyboardSpacer` and `KeyboardAwareScrollView` to work together. Both are relatively simple and I wanted full control over them because `react-native-keyboard-aware-scroll-view` wasn't a good fit for some of our uses cases and didn't handle platform specific edge cases very well.

Of course we can revisit and recheck all of this, but I wouldn't do this in this PR.

Happy to jump back on a conversation to detail this more.

Comment (etuleu) [reply packages/react-components/components/KeyboardAwareScrollView.tsx]: Ah, I see it's more complicated than I initially thought. I will revert this and put in a null check for the crash and then we can revisit this later. Thanks Jean!

**Review by jeanregisser — APPROVED** [review]
Android e2e test is green again!! w00t! 🏅 🚀

**Comment by etuleu** [comment]
> Android e2e test is green again!! w00t! medal_sports rocket

Amazing! thank you so much for your help @jeanregisser !

**Comment by aaronmgdr** [comment]
bumping this as it never merged

**Comment by Lss-Ankit** [comment]
Hi @etuleu @tarikbellamine We have performed e2e testing on latest Android internal build V1.7.2(1004294329) & Test Flight build V1.7.2(39) and we have observed that there are some crashes issue still are in open state. can you please look into those issues. 

**Test plan link:** https://docs.google.com/spreadsheets/d/1AFv9JWXnKfXJewhXf3nKlahB43d2v9xBU965ygXQ294/edit#gid=2098208089 

**Below are the crashes issues still are in open state**

1. https://github.com/celo-org/celo-monorepo/issues/4684
2. https://github.com/celo-org/celo-monorepo/issues/4685
3. https://github.com/celo-org/celo-monorepo/issues/6359
4. https://github.com/celo-org/celo-monorepo/issues/4704
